### ISSUE TITLE
Authentication Issue for external URL ( Power BI , .net Core 3.1, Azure AD App,IIS)

### ISSUE THREAD

**Issue by josephthomas5566** [AzureAD/microsoft-authentication-library-for-dotnet#3134]
We are building a **.net core app** to be hosted on **IIS**, which has an external/Public URL which converts into internal URL with app Proxy. The **.net core app is created as page within the main page.**

We have a **public** URL at : https://product.portal.example.org/link ,which internally via application proxy get redirected to **https://product.example.org/link.**

We have an AD App authentication which automatically reconstructs the return URI but uses "http://product.example.org/link" instead of "http://product.portal.example.org/link". Using custom redirect_uri we are able to redirect them to http://product.portal.example.org/link/signin-oidc.  But our issue is after the authentication it is not taking us to the  app page.

But within VPN , When the tester access the link https://product.example.org/link and logins (via AD APP) it come back with link on the URL https://product.example.org/link/signin-oidc and works fine. 

But outside  , When the tester access the link https://product.portal.example.org/link and logins (via AD APP) it come back with link on the URL https://product.portal.example.org/link/signin-oidc(GET). Then when it POST the final link : https://product.portal.example.org/link/home/app , ** it never succeeds. It continues to send the page to signin-oidc(get) and post page  ( Like a **LOOP** .. )

We are not sure what the issue is and has been stuck for some time. Any help is appreciated.

**Comment by jmprieur** [comment]
@josephthomas5566 : is it an ASP.NET Core app/
If yes, did you see https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-5.0#troubleshoot ?

**Comment by josephthomas5566** [comment]
yes . .net core app.. I will check this. Base code was taken from
https://github.com/microsoft/PowerBI-Developer-Samples/tree/master/.NET%20Core/Embed%20for%20your%20organization/UserOwnsData

To this we added custom redirect_uri

**Comment by josephthomas5566** [comment]
Unfortunately it is not working. Not sure if the redirect code is correct. This is still looping ( or not sure if the authentication is happening..) .. The Azure AD app has the return URI configured.. Anythoughts

          services.Configure<OpenIdConnectOptions>(OpenIdConnectDefaults.AuthenticationScheme, options =>
                {
                    options.SaveTokens = true; // this saves the token for the downstream api
                    options.Events = new OpenIdConnectEvents
                    {
                        OnRedirectToIdentityProvider = async ctxt =>
                        {

                            ctxt.ProtocolMessage.RedirectUri = "https://product.example.org/link/signin-oidc ";
                            await Task.Yield();
                        }
                    };
                });

**Comment by jmprieur** [comment]
@josephthomas5566, do you use Microsoft.Identity.Web?

**Comment by josephthomas5566** [comment]
Yes.. using Both. 
  using Microsoft.Identity.Web;
    using Microsoft.Identity.Web.UI;

Another observation is authentication is not returning <clientinfo> when we do redirect_uri ,but gives that when we do without that( on Internal- non public URL)
